deepmd/pt/model/descriptor/repformer_layer.py

```diff
@@ -1,125 +1,324 @@
+    precision
+        The precision of the residual tensor.
+    """
+    residual = nn.Parameter(
+        data=torch.zeros(_dim, dtype=PRECISION_DICT[precision], device=env.DEVICE),
+        requires_grad=trainable,
     )
+    if _mode == "norm":
+        nn.init.normal_(residual.data, std=_scale)
+    elif _mode == "const":
+        nn.init.constant_(residual.data, val=_scale)
+    else:
+        raise RuntimeError(f"Unsupported initialization mode '{_mode}'!")
+    return residual
 
 
+# common ops
 def _make_nei_g1(
     g1_ext: torch.Tensor,
     nlist: torch.Tensor,
 ) -> torch.Tensor:
-    # nlist: nb x nloc x nnei
-    nb, nloc, nnei = nlist.shape
-    # g1_ext: nb x nall x ng1
+    """
+    Make neighbor-wise atomic invariant rep.
+
+    Parameters
+    ----------
+    g1_ext
+        Extended atomic invariant rep, with shape nf x nall x ng1.
+    nlist
+        Neighbor list, with shape nf x nloc x nnei.
+
+    Returns
+    -------
+    gg1: torch.Tensor
+        Neighbor-wise atomic invariant rep, with shape nf x nloc x nnei x ng1.
+
+    """
+    # nlist: nf x nloc x nnei
+    nf, nloc, nnei = nlist.shape
+    # g1_ext: nf x nall x ng1
     ng1 = g1_ext.shape[-1]
-    # index: nb x (nloc x nnei) x ng1
-    index = nlist.reshape(nb, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
-    # gg1  : nb x (nloc x nnei) x ng1
+    # index: nf x (nloc x nnei) x ng1
+    index = nlist.reshape(nf, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
+    # gg1  : nf x (nloc x nnei) x ng1
     gg1 = torch.gather(g1_ext, dim=1, index=index)
-    # gg1  : nb x nloc x nnei x ng1
-    gg1 = gg1.view(nb, nloc, nnei, ng1)
+    # gg1  : nf x nloc x nnei x ng1
+    gg1 = gg1.view(nf, nloc, nnei, ng1)
     return gg1
 
 
 def _apply_nlist_mask(
     gg: torch.Tensor,
     nlist_mask: torch.Tensor,
 ) -> torch.Tensor:
-    # gg:  nf x nloc x nnei x ng
+    """
+    Apply nlist mask to neighbor-wise rep tensors.
+
+    Parameters
+    ----------
+    gg
+        Neighbor-wise rep tensors, with shape nf x nloc x nnei x d.
+    nlist_mask
+        Neighbor list mask, where zero means no neighbor, with shape nf x nloc x nnei.
+    """
+    # gg:  nf x nloc x nnei x d
     # msk: nf x nloc x nnei
     return gg.masked_fill(~nlist_mask.unsqueeze(-1), 0.0)
 
 
 def _apply_switch(gg: torch.Tensor, sw: torch.Tensor) -> torch.Tensor:
-    # gg:  nf x nloc x nnei x ng
+    """
+    Apply switch function to neighbor-wise rep tensors.
+
+    Parameters
+    ----------
+    gg
+        Neighbor-wise rep tensors, with shape nf x nloc x nnei x d.
+    sw
+        The switch function, which equals 1 within the rcut_smth range, smoothly decays from 1 to 0 between rcut_smth and rcut,
+        and remains 0 beyond rcut, with shape nf x nloc x nnei.
+    """
+    # gg:  nf x nloc x nnei x d
     # sw:  nf x nloc x nnei
     return gg * sw.unsqueeze(-1)
 
 
-def _apply_h_norm(
-    hh: torch.Tensor,  # nf x nloc x nnei x 3
+def _cal_hg(
```

Comment:
why renaming cal_h2g2 to cal_hg? any reason?

Reply:
Both update_g1_grrg and update_g1_drrd used this op (now in symmetry_op), maybe a more general name needed.